### PR TITLE
Fix TypeError when duplicating uncategorized theme patterns

### LINKED ISSUE #66888: DataViews: Duplicating an uncategorized pattern causes a JavaScript error
### Description

Originally reported on Trac: https://core.trac.wordpress.org/ticket/62367

> crash when duplicating "uncategorized" patterns
> condition: Fresh install of wordpress 6.7.RC4. Default 2025 theme.
> Recreate problem:

> Navigate "appearance" -> "editor" -> "patterns" -> "uncategorized"
> select one of the three patterns -> hamburger -> "duplicate"
> problem occurs
> symptom: black screen - stall -- needs browser refresh to restore.
> also throws javascript error;
> Uncaught TypeError: e.categories is undefined ... patterns.min.js:2:6657

I am able to reproduce this problem on WordPress 6.7-RC4-59375 with any theme that registers patterns without the `categories` parameter.

I tested this by removing the categories parameter from a pattern in Twenty Twenty-One, which is a classic theme that registers patterns with the PHP function, and in Twenty Twenty-Four, a block theme that registers patterns by using the `patterns` folder.
With and without Gutenberg trunk active.

```
react-dom.min.js?ver=18:1 Uncaught TypeError: Cannot read properties of undefined (reading 'includes')
    at duplicate-pattern-modal.js:20:24
    at Array.filter (<anonymous>)
    at L (duplicate-pattern-modal.js:19:6)
    at U (duplicate-pattern-modal.js:48:22)
    at RenderModal (duplicate-pattern.tsx:26:27)
    at gt (react-dom.min.js?ver=18:1:45667)
    at js (react-dom.min.js?ver=18:1:120123)
    at Sl (react-dom.min.js?ver=18:1:88331)
    at kl (react-dom.min.js?ver=18:1:88259)
    at bl (react-dom.min.js?ver=18:1:88122)
```

The `categories` parameter is not required for registering a pattern, and this should not cause an error.
https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#register_block_pattern

### Step-by-step reproduction instructions

> Navigate "appearance" -> "editor" -> "patterns" -> "uncategorized"
> select one of the three patterns -> hamburger -> "duplicate"
> problem occurs
> symptom: black screen - stall -- needs browser refresh to restore.
> also throws javascript error;
> Uncaught TypeError: e.categories is undefined ... patterns.min.js:2:6657

### Screenshots, screen recording, code snippet

_No response_

### Environment info

WordPress 6.7-RC4-59375

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [x] Block
- [x] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by carolinan**
## What Why & How

When a pattern is duplicated, there is a modal with an input field that contains the labels of the pattern categories:
the information about the category is passed on to the new pattern.

The function that gets the labels compares the category from the theme pattern with the existing registered pattern categories, to make sure that the category exists and is displayed with the correct label.

However, the code did not take into account that theme patterns can be registered without the optional `categories` parameter.
This PR makes a small change to the filter and mapping using optional chaining, to avoid a TypeError when the `categories` parameter is not used.

Closes https://github.com/WordPress/gutenberg/issues/66888

## Testing Instructions
Activate a theme that registers a pattern without the `categories` parameter. This can be a classic theme or a block theme, for example Twenty Twenty-Five.
Open the Patterns screen from Appearance > Patterns or from Appearance > Editor > Patterns.
Locate the pattern that does not have a category.
Select the option to duplicate the pattern.
Confirm that the modal with the information about the duplicated pattern opens and that there are no JavaScript errors in the browser console.
Click the button at the bottom of the modal to duplicate the pattern. Confirm that there are no issues related to the category of the new pattern.

Now, select a different pattern that has a single category or multiple categories, and duplicate it.
Confirm that the category labels show correctly in the modal and that there are no new issues/ regressions.
Click the button at the bottom of the modal to duplicate the pattern. Confirm that there are no issues related to the category of the new pattern.

### CODE REVIEW

File: packages/patterns/src/components/duplicate-pattern-modal.js [hunk 1]
```diff
@@ -15,11 +15,13 @@ import { PATTERN_SYNC_TYPES, PATTERN_TYPES } from '../constants';
...
-			?.filter( ( category ) =>
-				pattern.categories.includes( category.name )
-			)
-			.map( ( category ) => category.label );
+		return (
+			categories.core
+				?.filter( ( category ) =>
+					pattern.categories?.includes( category.name )
+				)
+				.map( ( category ) => category.label ) || []
```
Comment (ntsekouras) [inline packages/patterns/src/components/duplicate-pattern-modal.js]: Do we also need the empty array fallback? It seems it's handled as default by `CreatePatternModalContents`.

**Review by ntsekouras — APPROVED** [review]
Left a small comment, but this looks good. Thank you!

File: packages/patterns/src/components/duplicate-pattern-modal.js (hunk 1 above)
Comment (carolinan) [reply packages/patterns/src/components/duplicate-pattern-modal.js]: Thank you, I tested it again manually and removing the fallback does not seem to cause any issues.

**Comment by carolinan** [comment]
I briefly looked into adding a Playwright test for this but I was not able to figure out if there was a way to test a theme provided pattern.
Is there any way to tell the DataViews that the pattern origin is the theme, using `requestUtils.createBlock` ?

**Comment by carolinan** [comment]
Oh, or I am supposed to get theme patterns by `await requestUtils.activateTheme( 'theme name here' );`

**Comment by carolinan** [comment]
So I think the test should wait until TT5 is released, so that it can be added to the gutenberg-test-themes.

**Comment by ntsekouras** [comment]
For tests we're using the empty theme, where we can add stuff, like templates, patterns etc.. Did you try adding one there?

**Comment by carolinan** [comment]
I did not: I am torn about that because then it wont be _empty_ 😅

**Comment by ntsekouras** [comment]
>then it wont be empty

I get it 😄 , but it's not entirely empty right now. Anyway.. This fix is a JS pretty strait forward one, and I believe it would be fine not to add a test.

**Comment by mcsf** [comment]
Five lines below the patch site, there's a similar filter, but accessing property `wp_pattern_category` instead. Shouldn't we add optional chaining there too?

```diff
.filter( ( category ) =>
-	pattern.wp_pattern_category.includes( category.id )
+	pattern.wp_pattern_category?.includes( category.id )
)
```

@carolinan

**Comment by carolinan** [comment]
@mcsf I left it because I was not able to reproduce any errors when testing with user created patterns.
Since it is an existing taxonomy, registered on the post type, I believe that if there are no categories, it would be an empty array?
But yes, _in theory_ there is a chance that someone would unregister the `wp_pattern_category` taxonomy from the `wp_block` post type.

**Comment by carolinan** [comment]
@mcsf @ntsekouras 
Are there any other scenarios than the above where `wp_pattern_category` would not exist?

I ran a quick test and if I unregister the `wp_block_pattern` taxonomy, the block editor and site editor breaks because the taxonomy is used in [mapUserPattern ](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/private-selectors.js#L324)

The duplication of user created patterns do not break. I think it is something that needs to be looked at separately.

**Comment by cbravobernal** [comment]
I just cherry-picked this PR to the wp/6.7 branch to get it included in the next release.

**Pull request merged**